llvm/lib/CodeGen/MachineTraceMetrics.cpp

```diff
@@ -761,6 +762,64 @@ static void updatePhysDepsDownwards(const MachineInstr *UseMI,
   }
 }
 
+/// Estimates the number of cycles elapsed between DefMI and UseMI if they're
+/// non-null and in the same BasicBlock. Returns std::nullopt when UseMI is in a
+/// different MBB than DefMI.
+static std::optional<unsigned>
+estimateDefUseCycles(const TargetSchedModel &Sched, const MachineInstr *DefMI,
+                     const MachineInstr *UseMI) {
+  if (!DefMI || !UseMI || DefMI == UseMI)
+    return 0;
+  const MachineBasicBlock *ParentBB = DefMI->getParent();
+  if (ParentBB != UseMI->getParent())
+    return std::nullopt;
+
+  const auto DefIt =
+      llvm::find_if(ParentBB->instrs(),
+                    [DefMI](const MachineInstr &MI) { return DefMI == &MI; });
+  const auto UseIt =
+      llvm::find_if(ParentBB->instrs(),
+                    [UseMI](const MachineInstr &MI) { return UseMI == &MI; });
+
+  unsigned NumMicroOps = 0;
+  for (auto It = DefIt; It != UseIt; ++It) {
+    // In cases where the UseMI is a PHI at the beginning of the MBB, compute
+    // MicroOps until the end of the MBB.
+    if (It.isEnd())
+      break;
+
+    NumMicroOps += Sched.getNumMicroOps(&*It);
+  }
+  return NumMicroOps / Sched.getIssueWidth();
```

Comment:
I wonder if it would be useful to use `defaultDefLatency` for the instructions we are iterating over here in the case that some of the instructions we are iterating over depend on eachother. Imagine a scenario:

```
defmi = ...
a = ...
b = use a
usemi = ...
```
In this case, what if we checked defaultDefLatency of a and used it to understand the number of cycles elapsed between a and b? For example, if `a` has a default latency of 10, then b can't really start in the next `NumMicroOps / IssueWidth` cycles, since it has to wait 10 additional cycles.

Reply:
Sorry, I'm having trouble understanding your example. Here's what I think I understand:

```
c = ...
a = ...
b = use a
d = use c
```

Here, if c has a default def latency of N cycles, a has that of M cycles, b can start after M - 1 cycles, and d can start after N - 3 cycles, assuming issue-width = num-micro-ops = 1. What do you mean by "instructions that depends on each other"? How can one instruction depend on another, which in turn depends on the first? Wouldn't this break basic dominance criteria? Also, if I understand correctly, I think we're in SSA form at this point, so we don't have to worry about re-definitions.

Reply:
We have a loop that iterates over instructions between `DefMI` and `UseMI`:

```
for (auto It = DefIt; It != UseIt; ++It) {
    // In cases where the UseMI is a PHI at the beginning of the MBB, compute
    // MicroOps until the end of the MBB.
    if (It.isEnd())
      break;

    NumMicroOps += Sched.getNumMicroOps(&*It);
  }
```

I suggest the following scenario:
```
defmi = ...
a = ...
b = use a
usemi = ...
```
In this case, we will be looping over instructions `a` and `b` and adding their number of micro ops to calculate the number of cycles elapsed between `defmi` and `usemi`.

Let's take the assumption that `a` has default def latency of N cycles and `b` has default latency of M cycles.

> What do you mean by "instructions that depends on each other"?

In my scenario, `b` uses the result of `a`. It cannot start until that result is ready. If we want to make it concrete, we could imagine that it looks like this:

```
a = add 3, 2
b = sub a, 2
```

We cannot start the subtraction until `a` is finished calculating. This is what I am calling a dependency. We can assume for sake of simplicity here that `a` and `b` are independent from `defmi` and `usemi`.

In this scenario, I am suggesting that if the default latency of a is larger than the number of micro-ops, then we must wait at least the default latency of `a` before starting `b`. I suggest that we can incorporate this into the estimation.

Reply:
Ah, thanks for the explanation! However, I think we're missing the fact that CPUs are usually pipelined, and the defmi-usemi dependency will be in one pipeline, while the a-b dependency will be in another pipeline. Hence, I think the defmi-usemi dependency should be independent of the a-b dependency. When the code is called with DefMI = a, and UseMI = b, it will return the correct answer for that dependency.

Now, we haven't actually modeled any pipelines, but do you think this is feasible?

Reply:
It is a good point that if there are multiple pipelines and the defmi-usemi instructions goes down one and the a-b goes down another, then what you have is correct. There definitely is also a scenario where these dependency pairs go down the same pipeline and in that case what I am suggesting is probably the better model.

Unfortunately we don't have any pipeline information because we don't have the scheduler model, so we don't actually know what we should do.

One argument is to keep what we have no because its simple and less expensive to compute.
Another argument is to pick the "more common" approach. I'd prefer not to make a blanket statement and say that "it is more likely for independent instructions to go down different pipelines", although I wouldn't be surprised if this was the case.

For these reasons, I am content with the approach you are proposing.

Reply:
I may have missed something but the use of microops is worrying me - many arch don't guarantee that uop and latency are a close match (e.g. alderlake divpd ymm uops=1 latency=15).  And then dividing by issuewidth makes it feel more like a throughput estimate than latency?

Reply:
I see. Will adderlake divpd not dispatch in one cycle though? We're subtracting DefaultDefLatency by number of cycles elapsed between the DefMI and UseMI. If there's a adderlake divpd between the DefMI and UseMI, should we be subtracting 15 cycles for it?

Reply:
@RKSimon do you still have concerns about use of micro-ops and issue width here?

I think it makes sense to use issue width in determining "the number of cycles between two instructions" in the calculation here. For example:

```
a = def
b = ...
c = use
```

If the issue width is 1, then `a` is issued in once cycle, and `b` the next. But if the issue width is `2` then `a` and `b` are issued in the same cycle. In the former case we should estimate 2 cycles between `[a, c)` and in the latter estimate 1 cycle.

Reply:
Doesn't this makes assumptions that the ops can be issued on any pipe? I'm still not convinced your approach makes sense tbh.

Reply:
Since we don't have the scheduling model, we could either assume that all ops are issued on the same pipeline, or that independent ops are always issued on different pipelines. I think the latter case is more common. The patch is a first-order improvement over DefaultDefLatency: it is by no means an accurate representation of how the machine functions, but wouldn't you agree that the patch is an improvement over DefaultDefLatency?

Of course, if you have a better idea concerning how to improve to latency computation in the fallback case, please do suggest.